# **1.) Apriori algorithm implemented with the apyori package for Market Basket Optimization.**

In [28]:
#importing required packages
import pandas as pd
import numpy as np
!pip install apyori
from apyori import apriori

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [29]:
#loading market basket dataset..

df = pd.read_csv('data-1.csv',header=None)
df.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [30]:
# replacing empty value with 0. (Data cleaning)
df.fillna(0,inplace=True)
df.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs                 0             0   
2        chutney          0           0                 0             0   
3         turkey    avocado           0                 0             0   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1                 0     0               0             0             0   
2                 0     0               0             0             0   
3                 0     0               0             0             0   
4                 0     0               0             0             0   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1               0          0      0      0              0       0   
2               0          0      0      0              0       0   
3               0          0      0      0              0       0   
4               0          0      0      0              0       0   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                  0                0        0          0  
2                  0                0        0          0  
3                  0                0        0          0  
4                  0                0        0          0

In [31]:
# Data Pre-processing
# Converting data in list format to for using aprori

transactions = []
for i in range(0,len(df)):
    transactions.append([str(df.values[i,j]) for j in range(0,20) if str(df.values[i,j])!='0'])
transactions[0:3]

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney']]

In [32]:
# Calling apriori function
rules = apriori(transactions, min_support=0.003, min_confidance=0.2, min_lift=3, min_length=2)
rules
Results = list(rules)
## min_support = 0.003 -> means selecting items with min support of 0.3%
## min_confidance = 0.2 -> means min confidance of 20% 
## min_lift = 3  
## min_length = 2 -> means no. of items in the transaction should be 2

In [33]:
# creating dataframe of the rules
df_results = pd.DataFrame(Results)
support = df_results.support
df_results.head()

items   support  \
0        (cottage cheese, brownies)  0.003466   
1            (chicken, light cream)  0.004533   
2  (escalope, mushroom cream sauce)  0.005733   
3                 (escalope, pasta)  0.005866   
4       (tomato juice, fresh bread)  0.004266   

                                  ordered_statistics  
0  [((brownies), (cottage cheese), 0.102766798418...  
1  [((chicken), (light cream), 0.0755555555555555...  
2  [((escalope), (mushroom cream sauce), 0.072268...  
3  [((escalope), (pasta), 0.07394957983193277, 4....  
4  [((fresh bread), (tomato juice), 0.09907120743...

In [34]:
#Creating 4 list for optimization
first_values = []
second_values = []
third_values = []
fourth_value = []

for i in range(df_results.shape[0]):
    single_list = df_results['ordered_statistics'][i][0]
    first_values.append(list(single_list[0]))
    second_values.append(list(single_list[1]))
    third_values.append(single_list[2])
    fourth_value.append(single_list[3])

In [35]:
# creating data frame of all lists
lhs = pd.DataFrame(first_values)
rhs = pd.DataFrame(second_values)
confidance=pd.DataFrame(third_values,columns=['Confidance'])
lift=pd.DataFrame(fourth_value,columns=['lift'])

In [36]:
# concat all list together in a single dataframe
df_final = pd.concat([lhs,rhs,support,confidance,lift], axis=1)
df_final.head()

0     1                     0     1     2   support  Confidance  \
0     brownies  None        cottage cheese  None  None  0.003466    0.102767   
1      chicken  None           light cream  None  None  0.004533    0.075556   
2     escalope  None  mushroom cream sauce  None  None  0.005733    0.072269   
3     escalope  None                 pasta  None  None  0.005866    0.073950   
4  fresh bread  None          tomato juice  None  None  0.004266    0.099071   

       lift  
0  3.225330  
1  4.843951  
2  3.790833  
3  4.700812  
4  3.259356

In [37]:
df_final.fillna(value=' ', inplace=True)
df_final.head()

0  1                     0  1  2   support  Confidance      lift
0     brownies           cottage cheese        0.003466    0.102767  3.225330
1      chicken              light cream        0.004533    0.075556  4.843951
2     escalope     mushroom cream sauce        0.005733    0.072269  3.790833
3     escalope                    pasta        0.005866    0.073950  4.700812
4  fresh bread             tomato juice        0.004266    0.099071  3.259356

In [38]:
#set column name
df_final.columns = ['lhs',1,'rhs',2,3,'support','confidance','lift']
df_final.head()   

lhs  1                   rhs  2  3   support  confidance      lift
0     brownies           cottage cheese        0.003466    0.102767  3.225330
1      chicken              light cream        0.004533    0.075556  4.843951
2     escalope     mushroom cream sauce        0.005733    0.072269  3.790833
3     escalope                    pasta        0.005866    0.073950  4.700812
4  fresh bread             tomato juice        0.004266    0.099071  3.259356

In [39]:
# Adding column 1 in lhs and 2,3 in rhs then dropping them
df_final['lhs'] = df_final['lhs'] + str(", ") + df_final[1]
df_final['rhs'] = df_final['rhs']+str(", ")+df_final[2] + str(", ") + df_final[3]
df_final.drop(columns=[1,2,3],inplace=True)
df_final.head()

lhs                         rhs   support  confidance      lift
0     brownies,          cottage cheese,  ,    0.003466    0.102767  3.225330
1      chicken,             light cream,  ,    0.004533    0.075556  4.843951
2     escalope,    mushroom cream sauce,  ,    0.005733    0.072269  3.790833
3     escalope,                   pasta,  ,    0.005866    0.073950  4.700812
4  fresh bread,            tomato juice,  ,    0.004266    0.099071  3.259356

In [40]:
#Displaying the top 5 values after sorting by lift
df_final.sort_values('lift', ascending=False).head()

lhs                                  rhs   support  \
58      olive oil,    mineral water, whole wheat pasta,    0.003866   
6   fromage blanc,                            honey,  ,    0.003333   
49    ground beef,             tomato sauce, spaghetti,    0.003066   
1         chicken,                      light cream,  ,    0.004533   
3        escalope,                            pasta,  ,    0.005866   

    confidance      lift  
58    0.058704  6.115863  
6     0.245098  5.164271  
49    0.031208  4.980600  
1     0.075556  4.843951  
3     0.073950  4.700812

# **2.) Apriori algorithm implemented for Market Basket Optimization using the mlxtend.frequent_patterns package.**

Why are we using this method?

1. The first approach had a limitation of requiring data conversion into a list format, which can be computationally expensive for stores with a large number of transactions.

2. The Apyori package is outdated and may not offer the latest functionalities.

3. The results obtained from the previous approach are presented in a less user-friendly format, indicating the need for additional pre-processing steps.

4. Instead, we can utilize the mlxtend package, which not only generates frequent itemsets but also provides association rules, addressing the limitations mentioned earlier.

In [41]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

df1 = pd.read_csv('data-2.csv', encoding="ISO-8859-1")
df1.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [42]:
df1.Country.value_counts()

United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

In [43]:
df1 = df1[df1.Country == 'France'] # Since data has many country choosing any one for check.

In [44]:
df1['Description'] = df1['Description'].str.strip() # Remoiving spaces
df1 = df1[df1.Quantity >0] # removing transcation with neagtive values
df1.head()

<ipython-input-44-ff8fddc87e79>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Description'] = df1['Description'].str.strip() # Remoiving spaces


InvoiceNo StockCode                      Description  Quantity  \
26    536370     22728        ALARM CLOCK BAKELIKE PINK        24   
27    536370     22727         ALARM CLOCK BAKELIKE RED        24   
28    536370     22726       ALARM CLOCK BAKELIKE GREEN        12   
29    536370     21724  PANDA AND BUNNIES STICKER SHEET        12   
30    536370     21883                  STARS GIFT TAPE        24   

       InvoiceDate  UnitPrice  CustomerID Country  
26  12/1/2010 8:45       3.75     12583.0  France  
27  12/1/2010 8:45       3.75     12583.0  France  
28  12/1/2010 8:45       3.75     12583.0  France  
29  12/1/2010 8:45       0.85     12583.0  France  
30  12/1/2010 8:45       0.65     12583.0  France

In [45]:
# creating pivot table with Quantity sum as values and filling 0 if any nan values

basket = pd.pivot_table(data=df1,index='InvoiceNo',columns='Description',values='Quantity', aggfunc='sum',fill_value=0)
basket.head()

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536370                            0                           0   
536852                            0                           0   
536974                            0                           0   
537065                            0                           0   
537463                            0                           0   

Description  12 EGG HOUSE PAINTED WOOD  12 MESSAGE CARDS WITH ENVELOPES  \
InvoiceNo                                                                 
536370                               0                                0   
536852                               0                                0   
536974                               0                                0   
537065                               0                                0   
537463                               0                                0   

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
536370                                   0   
536852                                   0   
536974                                   0   
537065                                   0   
537463                                   0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536370                                         0                            0   
536852                                         0                            0   
536974                                         0                            0   
537065                                         0                            0   
537463                                         0                            0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536370                               0                                   0   
536852                               0                                   0   
536974                               0                                   0   
537065                               0                                   0   
537463                               0                                   0   

Description  12 PENCILS TALL TUBE WOODLAND  ...  WRAP VINTAGE PETALS  DESIGN  \
InvoiceNo                                   ...                                
536370                                   0  ...                            0   
536852                                   0  ...                            0   
536974                                   0  ...                            0   
537065                                   0  ...                            0   
537463                                   0  ...                            0   

Description  YELLOW COAT RACK PARIS FASHION  YELLOW GIANT GARDEN THERMOMETER  \
InvoiceNo                                                                      
536370                                    0                                0   
536852                                    0                                0   
536974                                    0                                0   
537065                                    0                                0   
537463                                    0                                0   

Description  YELLOW SHARK HELICOPTER  ZINC  STAR T-LIGHT HOLDER  \
InvoiceNo                                                         
536370                             0                          0   
536852                             0                          0   
536974                             0                          0   
537065                             0                          0   
537463                             0      

In [46]:
# checking correctness after binning to 1 
basket['10 COLOUR SPACEBOY PEN'].head(6)

InvoiceNo
536370     0
536852     0
536974     0
537065     0
537463     0
537468    24
Name: 10 COLOUR SPACEBOY PEN, dtype: int64

In [47]:
# converting to 1 if more then 0 else o
def convert_into_binary(x):
    if x > 0:
        return 1
    else:
        return 0
basket_sets = basket.applymap(convert_into_binary)
basket_sets['10 COLOUR SPACEBOY PEN'].head(6)

InvoiceNo
536370    0
536852    0
536974    0
537065    0
537463    0
537468    1
Name: 10 COLOUR SPACEBOY PEN, dtype: int64

In [48]:
# removing postage since its insignificant
basket_sets.drop(columns=['POSTAGE'],inplace=True)

In [49]:
# calling apriori function and passing minimum support.
frequent_itemsets = apriori(basket_sets, min_support=0.07, use_colnames=True)
frequent_itemsets.head()

support                        itemsets
0  0.071429   (4 TRADITIONAL SPINNING TOPS)
1  0.096939    (ALARM CLOCK BAKELIKE GREEN)
2  0.102041     (ALARM CLOCK BAKELIKE PINK)
3  0.094388      (ALARM CLOCK BAKELIKE RED)
4  0.081633  (BAKING SET 9 PIECE RETROSPOT)

In [50]:
# setting metric as lift and keeping minimum lift as 1
rules_mlxtend = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules_mlxtend.head()

antecedents                   consequents  \
0  (ALARM CLOCK BAKELIKE GREEN)   (ALARM CLOCK BAKELIKE PINK)   
1   (ALARM CLOCK BAKELIKE PINK)  (ALARM CLOCK BAKELIKE GREEN)   
2  (ALARM CLOCK BAKELIKE GREEN)    (ALARM CLOCK BAKELIKE RED)   
3    (ALARM CLOCK BAKELIKE RED)  (ALARM CLOCK BAKELIKE GREEN)   
4   (ALARM CLOCK BAKELIKE PINK)    (ALARM CLOCK BAKELIKE RED)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.096939            0.102041  0.073980    0.763158  7.478947   
1            0.102041            0.096939  0.073980    0.725000  7.478947   
2            0.096939            0.094388  0.079082    0.815789  8.642959   
3            0.094388            0.096939  0.079082    0.837838  8.642959   
4            0.102041            0.094388  0.073980    0.725000  7.681081   

   leverage  conviction  
0  0.064088    3.791383  
1  0.064088    3.283859  
2  0.069932    4.916181  
3  0.069932    5.568878  
4  0.064348    3.293135

In [51]:
# Displaying items with lift >= 4 and confidence >= 0.8
rules_mlxtend[ (rules_mlxtend['lift'] >= 4) & (rules_mlxtend['confidence'] >= 0.8) ]

antecedents  \
2                        (ALARM CLOCK BAKELIKE GREEN)   
3                          (ALARM CLOCK BAKELIKE RED)   
16                    (SET/6 RED SPOTTY PAPER PLATES)   
18                    (SET/6 RED SPOTTY PAPER PLATES)   
19                      (SET/6 RED SPOTTY PAPER CUPS)   
20  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...   
21  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...   
22  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   

                             consequents  antecedent support  \
2             (ALARM CLOCK BAKELIKE RED)            0.096939   
3           (ALARM CLOCK BAKELIKE GREEN)            0.094388   
16  (SET/20 RED RETROSPOT PAPER NAPKINS)            0.127551   
18         (SET/6 RED SPOTTY PAPER CUPS)            0.127551   
19       (SET/6 RED SPOTTY PAPER PLATES)            0.137755   
20         (SET/6 RED SPOTTY PAPER CUPS)            0.102041   
21  (SET/20 RED RETROSPOT PAPER NAPKINS)            0.122449   
22       (SET/6 RED SPOTTY PAPER PLATES)            0.102041   

    consequent support   support  confidence      lift  leverage  conviction  
2             0.094388  0.079082    0.815789  8.642959  0.069932    4.916181  
3             0.096939  0.079082    0.837838  8.642959  0.069932    5.568878  
16            0.132653  0.102041    0.800000  6.030769  0.085121    4.336735  
18            0.137755  0.122449    0.960000  6.968889  0.104878   21.556122  
19            0.127551  0.122449    0.888889  6.968889  0.104878    7.852041  
20            0.137755  0.099490    0.975000  7.077778  0.085433   34.489796  
21            0.132653  0.099490    0.812500  6.125000  0.083247    4.625850  
22            0.127551  0.099490    0.975000  7.644000  0.086474   34.897959

1. Antecedent support: This measure indicates the frequency of the antecedent item in all the transactions. For example, the item "ALARM CLOCK BAKELIKE RED" is present in 9.4% of the transactions.

2. Consequent support: This measure indicates the frequency of the consequent item in all the transactions. For example, the item "ALARM CLOCK BAKELIKE GREEN" is present in 9.6% of the transactions.

3. Support: This measure indicates the frequency of the itemset in all the transactions. For example, the itemset {ALARM CLOCK BAKELIKE RED, ALARM CLOCK BAKELIKE GREEN} is present in 7.9% of the transactions.

4. Confidence: This measure defines the likelihood of finding the consequent item in the cart given that the cart already contains the antecedents. For example, {ALARM CLOCK BAKELIKE RED} -> {ALARM CLOCK BAKELIKE GREEN} has a confidence of 83%. In simple terms, there is an 83% chance of finding {ALARM CLOCK BAKELIKE GREEN} if the cart contains {ALARM CLOCK BAKELIKE RED}.

5. Lift: This measure defines the likelihood of finding the consequent item in the cart given that the cart already contains the antecedent, while considering the popularity of the consequent item. For example, the lift of {ALARM CLOCK BAKELIKE GREEN} with respect to {ALARM CLOCK BAKELIKE RED} is 8.64, indicating a strong association. A lift value greater than 1 implies that the association rule is worth considering.

6. Leverage: Leverage measures the difference between the occurrence of X and Y appearing together in the dataset and what would be expected if X and Y were statistically independent. In a sales setting, it helps determine the additional units (items X and Y together) sold compared to what would be expected from independent sales.

7. Conviction: This parameter is not much useful in most situations and can be ignored.